core/src/main/java/org/apache/iceberg/rest/TableScanResponseParser.java

```diff
@@ -77,7 +77,17 @@ public static List<FileScanTask> parseFileScanTasks(
         fileScanTaskList.add(fileScanTask);
       }
 
+      if (fileScanTaskList.isEmpty()) {
+        Preconditions.checkArgument(
+            (deleteFiles == null || deleteFiles.isEmpty()),
+            "Invalid response: deleteFiles should only be returned with fileScanTasks that reference them");
+      }
+
       return fileScanTaskList;
+    } else {
+      Preconditions.checkArgument(
+          (deleteFiles == null || deleteFiles.isEmpty()),
+          "Invalid response: deleteFiles should only be returned with fileScanTasks that reference them");
```

Comment:
I think this is the real place to do this check. It makes sense to validate on parsing JSON coming over the wire. It does not make sense to do this check after building the response because our builder shouldn't even allow that situation to happen.

Reply:
Once we properly remove the API in 1.12, we can remove this check from all the places it's done in validate().